### PR TITLE
feat: pdf crafter

### PR DESCRIPTION
**Pull request by Yongxuanzhang**
https://github.com/jina-ai/internal-tasks/issues/92

### CODE REVIEW

File: pdf_crafter.py [hunk 1]
```diff
@@ -1,31 +1,93 @@
...
+        pdf_text = None
+        try:
+            if doc.uri:
+                pdf_img = fitz.open(doc.uri)
+                pdf_text = pdfplumber.open(doc.uri)
+            if doc.buffer:
+                pdf_img = fitz.open(stream=doc.buffer, filetype='pdf')
+                pdf_text = pdfplumber.open(io.BytesIO(doc.buffer))
+        except Exception as ex:
+            logger.error(f'Failed to open due to: {ex}')
```
Comment (JoanFM) [inline pdf_crafter.py]: instantiate new JinaLogger, do not use default logger. Core Jina has nothing to do with these logs

**Review by JoanFM — CHANGES_REQUESTED** [review]

File: pdf_crafter.py [hunk 2]
```diff
@@ -1,31 +1,93 @@
...
\ No newline at end of file
+            pdf_img, pdf_text = self._parse_pdf(doc)
+
+            if pdf_img is not None:
+                images = self._extract_image(pdf_img)
+                doc.chunks += [Document(blob=img, mime_type='image/*') for img in images]
+                self._tag_with_root_doc_id(doc, level='c')
+            if pdf_text is not None:
+                texts = self._extract_text(pdf_text)
+                doc.chunks += [Document(text=t, mime_type='text/plain') for t in texts]
```
Comment (JoanFM) [inline pdf_crafter.py]: doc chunks append

**Review by JoanFM — DISMISSED** [review]

File: pdf_crafter.py (hunk 1 above)
Comment (Yongxuanzhang) [reply pdf_crafter.py]: how about now?

File: pdf_crafter.py (hunk 2 above)
Comment (Yongxuanzhang) [reply pdf_crafter.py]: fixed

File: README.md [hunk 3]
```diff
@@ -10,7 +10,7 @@ PDFPlumberSegmenter is a segmenter used for extracting images and text as chunks
 - [🌱 Prerequisites](#-prerequisites)
 - [🚀 Usages](#-usages)
 - [🎉️ Example](#%EF%B8%8F-example)
-- [🔍️ Reference](#%EF%B8%8F-reference)
+
```
Comment (CatStark) [inline README.md]: What is that "%EF%B8%8F" in the example?

File: README.md [hunk 4]
```diff
@@ -46,34 +46,34 @@ Use the source codes from JinaHub in your python codes,
...
-    uses: 'jinahub://MyDummyExecutor'
+  - name: crafter
+    uses: 'jinahub://PDFCrafter'
 ```
 
 
 ### 📦️ Via Pypi
 
-1. Install the `jinahub-MY-DUMMY-EXECUTOR` package.
+1. Install the `executor-pdf-crafterR` package.
```
Comment (CatStark) [inline README.md]: Last letter should be lower case right?

File: README.md [hunk 5]
```diff
@@ -28,16 +28,16 @@ Use the prebuilt images from JinaHub in your python codes,
...
 ```yaml
 jtype: Flow
 pods:
-  - name: encoder
-    uses: 'jinahub+docker://MyDummyExecutor'
+  - name: crafter
+    uses: 'jinahub+docker://PDFCrafter'
 ```
 
 #### using source codes
```
Comment (slettner) [inline README.md:32]: The `using source codes` section looks very similar to the above two snippets. I would remove it, as I see it, it does not provide any good insights

File: pdf_crafter.py [hunk 6]
```diff
@@ -1,31 +1,102 @@
...
+    def craft(self, docs: DocumentArray, **kwargs):
         """
-        Encode all docs with images and store the encodings in the embedding attribute of the docs.
-        :param docs: documents sent to the encoder. The docs must have `blob` of the shape `256`.
+        Read PDF files. Extracts data from them.
+        Checks if the input is a string of the filename,
+        or if it's the file in bytes.
+        It will then extract the data from the file, creating a list for images,
+        and text.
+        :param docs: Array of Documents.
```
Comment (slettner) [inline pdf_crafter.py:39]:
```suggestion
        :param docs: DocumentArray containing images
        :param parameters: dictionary to define the `traversal_paths` and the `batch_size`. For example,
               `parameters={'traversal_paths': ['r'], 'batch_size': 10}`.
        :param kwargs: Additional key value arguments.
```

File: pdf_crafter.py [hunk 7]
```diff
@@ -1,31 +1,102 @@
...
 
 class PDFCrafter(Executor):
     """
-    Encode the Document blob into embedding.
-
-    :param embedding_dim: the output dimensionality of the embedding
+    :class:`PDFCrafter` Extracts data (text and images) from PDF files.
+    Stores images (`mime_type`=image/*) on chunk level ('c') and text segments (`mime_type`=text/plain)
+    on chunk level ('c') in the root ('r') Document.
+    Text is further split by linebreaks and stored on chunk-chunk level ('cc')
```
Comment (slettner) [inline pdf_crafter.py]: This is not happening in the Crafter. Not sure if we should add it

File: pdf_crafter.py [hunk 8]
```diff
@@ -1,31 +1,102 @@
...
+                        else:
+                            images.append(np_arr)
+                    else:  # if CMYK:
+                        pix = fitz.Pixmap(fitz.csRGB, pix)  # Convert to RGB
+                        np_arr = np.frombuffer(pix.samples, dtype=np.uint8).reshape(pix.h, pix.w, pix.n).astype(
+                            'float32')
+                        images.append(np_arr)
+        return images
+
+    def _tag_with_root_doc_id(self, doc: Document, level: str):
```
Comment (slettner) [inline pdf_crafter.py]: I would remove `_tag_with_root_doc_id`. Not really clear why we need this here (I know where it came from, but seeing the Crafter isolated it is hard to understand what the intention of this function is)

File: tests/integration/test_exec.py [hunk 9]
```diff
@@ -1,12 +1,45 @@
...
+from jina import Flow, Document, DocumentArray
 from jinahub.crafter.pdf_crafter import PDFCrafter
 
 
-def test_exec():
-    f = Flow().add(uses=MyDummyExecutor)
-    with f:
-        resp = f.post(on='/test', inputs=Document(), return_results=True)
-        assert resp is not None
+def doc_generator(uri, buffer):
```
Comment (slettner) [inline tests/integration/test_exec.py]: You could turn this into a fixture that returns a Generator. Maybe collapse it with the `input_pdf` fixture.
I can be used in the unit test as well then.

**Review by slettner — CHANGES_REQUESTED** [review]
The Crafter needs some changes I think, tests are looking good. Great Job! 
Let's get this done and merged.

File: README.md (hunk 3 above)
Comment (Yongxuanzhang) [reply README.md]: this is from the template, seems it could still work?

File: README.md (hunk 4 above)
Comment (Yongxuanzhang) [reply README.md]: yes, thanks!

File: README.md (hunk 5 above)
Comment (Yongxuanzhang) [reply README.md:32]: it's different I think. using source codes means pull the code and use the code, the user need to install requirements. Previous snippet is about using docker from hub

File: pdf_crafter.py (hunk 7 above)
Comment (Yongxuanzhang) [reply pdf_crafter.py]: yes removed. this is a copy-paste from our example repo. I should have checked the docstrings.

File: pdf_crafter.py (hunk 8 above)
Comment (Yongxuanzhang) [reply pdf_crafter.py]: ok I removed it

File: tests/integration/test_exec.py (hunk 9 above)
Comment (Yongxuanzhang) [reply tests/integration/test_exec.py]: good point, thanks

File: README.md (hunk 3 above)
Comment (nan-wang) [reply README.md]: It is safely to delete this part and let the `toc.yml` from the Github Actions to generate them automatically

**Review by nan-wang — DISMISSED** [review]

File: README.md (hunk 5 above)
Comment (slettner) [reply README.md:32]: Ah ok, I thought it was just another docker file used in the same way as from the hub.

**Review by slettner — APPROVED** [review]
Looks good from my side.

File: README.md (hunk 3 above)
Comment (Yongxuanzhang) [reply README.md]: sure ， that's awesome

**Pull request merged**